"use strict";(self.webpackChunkjob=self.webpackChunkjob||[]).push([[8708],{3905:(e,r,n)=>{n.d(r,{Zo:()=>u,kt:()=>h});var t=n(7294);function o(e,r,n){return r in e?Object.defineProperty(e,r,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[r]=n,e}function a(e,r){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var t=Object.getOwnPropertySymbols(e);r&&(t=t.filter((function(r){return Object.getOwnPropertyDescriptor(e,r).enumerable}))),n.push.apply(n,t)}return n}function i(e){for(var r=1;r<arguments.length;r++){var n=null!=arguments[r]?arguments[r]:{};r%2?a(Object(n),!0).forEach((function(r){o(e,r,n[r])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):a(Object(n)).forEach((function(r){Object.defineProperty(e,r,Object.getOwnPropertyDescriptor(n,r))}))}return e}function s(e,r){if(null==e)return{};var n,t,o=function(e,r){if(null==e)return{};var n,t,o={},a=Object.keys(e);for(t=0;t<a.length;t++)n=a[t],r.indexOf(n)>=0||(o[n]=e[n]);return o}(e,r);if(Object.getOwnPropertySymbols){var a=Object.getOwnPropertySymbols(e);for(t=0;t<a.length;t++)n=a[t],r.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(o[n]=e[n])}return o}var l=t.createContext({}),c=function(e){var r=t.useContext(l),n=r;return e&&(n="function"==typeof e?e(r):i(i({},r),e)),n},u=function(e){var r=c(e.components);return t.createElement(l.Provider,{value:r},e.children)},p="mdxType",d={inlineCode:"code",wrapper:function(e){var r=e.children;return t.createElement(t.Fragment,{},r)}},m=t.forwardRef((function(e,r){var n=e.components,o=e.mdxType,a=e.originalType,l=e.parentName,u=s(e,["components","mdxType","originalType","parentName"]),p=c(n),m=o,h=p["".concat(l,".").concat(m)]||p[m]||d[m]||a;return n?t.createElement(h,i(i({ref:r},u),{},{components:n})):t.createElement(h,i({ref:r},u))}));function h(e,r){var n=arguments,o=r&&r.mdxType;if("string"==typeof e||o){var a=n.length,i=new Array(a);i[0]=m;var s={};for(var l in r)hasOwnProperty.call(r,l)&&(s[l]=r[l]);s.originalType=e,s[p]="string"==typeof e?e:o,i[1]=s;for(var c=2;c<a;c++)i[c]=n[c];return t.createElement.apply(null,i)}return t.createElement.apply(null,n)}m.displayName="MDXCreateElement"},5009:(e,r,n)=>{n.r(r),n.d(r,{assets:()=>l,contentTitle:()=>i,default:()=>p,frontMatter:()=>a,metadata:()=>s,toc:()=>c});var t=n(7462),o=(n(7294),n(3905));const a={},i="How do you handle errors in a NodeJS application?",s={unversionedId:"questions/nodejs/nodejs-error-handling",id:"questions/nodejs/nodejs-error-handling",title:"How do you handle errors in a NodeJS application?",description:"An error is any problem given out by the program due to a number of factors such as logic, syntax, timeout, etc. An error in Node.js is any instance of the Error object. Common examples include built-in error classes, such as ReferenceError, RangeError, TypeError, URIError, EvalError, and SyntaxError.",source:"@site/interview/questions/nodejs/nodejs-error-handling.md",sourceDirName:"questions/nodejs",slug:"/questions/nodejs/nodejs-error-handling",permalink:"/questions/nodejs/nodejs-error-handling",draft:!1,editUrl:"https://github.com/izemil/job/edit/master/interview/questions/nodejs/nodejs-error-handling.md",tags:[],version:"current",frontMatter:{},sidebar:"tutorialSidebar",previous:{title:"Why/How to create a nodejs cluster?",permalink:"/questions/nodejs/nodejs-clustering"},next:{title:"What is an Event Loop?",permalink:"/questions/nodejs/nodejs-event-loop"}},l={},c=[{value:"Error handling",id:"error-handling",level:2}],u={toc:c};function p(e){let{components:r,...n}=e;return(0,o.kt)("wrapper",(0,t.Z)({},u,n,{components:r,mdxType:"MDXLayout"}),(0,o.kt)("h1",{id:"how-do-you-handle-errors-in-a-nodejs-application"},"How do you handle errors in a NodeJS application?"),(0,o.kt)("p",null,"An error is any problem given out by the program due to a number of factors such as logic, syntax, timeout, etc. An error in Node.js is any instance of the Error object. Common examples include built-in error classes, such as ReferenceError, RangeError, TypeError, URIError, EvalError, and SyntaxError."),(0,o.kt)("p",null,"User-defined errors can also be created by extending the base Error object, a built-in error class, or another custom error. In general, Node.js errors are divided into two distinct categories: operational errors and programmer errors."),(0,o.kt)("p",null,(0,o.kt)("strong",{parentName:"p"},"1. Operational Errors:")),(0,o.kt)("p",null,"Operational errors represent runtime problems. These errors are expected in the Node.js runtime and should be dealt with in a proper way. Here's a list of common operational errors:"),(0,o.kt)("ul",null,(0,o.kt)("li",{parentName:"ul"},"failed to connect to server"),(0,o.kt)("li",{parentName:"ul"},"failed to resolve hostname"),(0,o.kt)("li",{parentName:"ul"},"invalid user input"),(0,o.kt)("li",{parentName:"ul"},"request timeout"),(0,o.kt)("li",{parentName:"ul"},"server returned a 500 response"),(0,o.kt)("li",{parentName:"ul"},"socket hang-up"),(0,o.kt)("li",{parentName:"ul"},"system is out of memory")),(0,o.kt)("p",null,(0,o.kt)("strong",{parentName:"p"},"2. Programmer Errors:")),(0,o.kt)("p",null,"Programmer errors are what we call bugs. They represent issues in the code itself. Here's a common one for Node.js, when you try reading a property of an undefined object. It's a classic case of programmer error. Here are a few more:"),(0,o.kt)("ul",null,(0,o.kt)("li",{parentName:"ul"},"called an asynchronous function without a callback"),(0,o.kt)("li",{parentName:"ul"},"did not resolve a promise"),(0,o.kt)("li",{parentName:"ul"},"did not catch a rejected promise"),(0,o.kt)("li",{parentName:"ul"},"passed a string where an object was expected"),(0,o.kt)("li",{parentName:"ul"},"passed an object where a string was expected"),(0,o.kt)("li",{parentName:"ul"},"passed incorrect parameters in a function")),(0,o.kt)("h2",{id:"error-handling"},"Error handling"),(0,o.kt)("p",null,"In Node.js, errors can be handled using try-catch blocks or by listening for the 'error' event on an event emitter. It is also a good practice to use a global error-handling middleware function to catch unhandled errors."),(0,o.kt)("p",null,"Unhandled exceptions in Node.js can be caught at the Process level by attaching a handler for uncaughtException event."),(0,o.kt)("pre",null,(0,o.kt)("code",{parentName:"pre"},"process.on('uncaughtException', function(err) {\n    console.log('Caught exception: ' + err);\n});\n")),(0,o.kt)("p",null,"Process is a global object that provides information about the current Node.js process. Process is a listener function that is always listening to events."),(0,o.kt)("p",null,"Few events are :"),(0,o.kt)("ol",null,(0,o.kt)("li",{parentName:"ol"},(0,o.kt)("inlineCode",{parentName:"li"},"exit")),(0,o.kt)("li",{parentName:"ol"},(0,o.kt)("inlineCode",{parentName:"li"},"disconnect")),(0,o.kt)("li",{parentName:"ol"},(0,o.kt)("inlineCode",{parentName:"li"},"unhandledException")),(0,o.kt)("li",{parentName:"ol"},(0,o.kt)("inlineCode",{parentName:"li"},"rejectionHandled"))))}p.isMDXComponent=!0}}]);